tests/test_pythonpackage.py

```diff
@@ -0,0 +1,133 @@
+            "sys.path = [os.path.dirname(sys.argv[1])] + sys.path\n"
+            "m = importlib.import_module(\n"
+            "    os.path.basename(sys.argv[1]).partition('.')[0]\n"
+            ")\n"
+            "print(m._get_system_python_executable())",
+            os.path.join(os.path.dirname(__file__), "..",
+                         "pythonforandroid", "pythonpackage.py"),
+        ]
+        # Actual call to python:
+        try:
+            result = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            raise RuntimeError("call failed, with output: " + str(e.output))
+        try:
+            return result.decode("utf-8", "replace").strip()
+        except AttributeError:
+            return result.strip()
+
+    # Check that in system-wide python, the system-wide python is returned:
+    try:
+        p1 = os.path.normpath(run__get_system_python_executable(pybin))
+        p2 = os.path.normpath(pybin)
+        assert(p1 == p2)
+    except RuntimeError as e:
+        if "pep517" in str(e.args):
+            # System python probably doesn't have pep517 available.
+            # Not much we can do in that case.
+            pass
+        else:
+            raise e
+
+    test_dir = tempfile.mkdtemp()
+    try:
+        # Check that in a virtualenv, the system-wide python is returned:
+        subprocess.check_output([
+            pybin, "-m", "virtualenv",
+            "--python=" + str(sys.executable),
+            "--",
+            os.path.join(test_dir, "virtualenv")
+        ])
+        subprocess.check_output([
+            os.path.join(test_dir, "virtualenv", "bin", "pip"),
+            "install", "-U", "pip"
+        ])
+        subprocess.check_output([
+            os.path.join(test_dir, "virtualenv", "bin", "pip"),
+            "install", "-U", "pep517"
+        ])
+        assert(os.path.normpath(run__get_system_python_executable(
+                                os.path.join(test_dir, "virtualenv",
+                                             "bin", "python"))) ==
+               os.path.normpath(pybin))
+        if int(sys.version.partition(".")[0]) >= 3:
+            # Check that in a venv/pyvenv, the system-wide python is returned:
+            subprocess.check_output([
+                pybin, "-m", "venv", "--",
+                os.path.join(test_dir, "venv")
+            ])
+            subprocess.check_output([
+                os.path.join(test_dir, "venv", "bin", "pip"),
+                "install", "-U", "pip"
+            ])
+            subprocess.check_output([
+                os.path.join(test_dir, "venv", "bin", "pip"),
+                "install", "-U", "pep517"
+            ])
+            assert(os.path.normpath(run__get_system_python_executable(
+                                    os.path.join(test_dir, "virtualenv",
+                                                 "bin", "python"))) ==
+                   os.path.normpath(pybin))
```

Comment:
It's pretty hard to say what we're testing and what we're not testing here. It's definitely more an integration test here going as far as touching the filesystem, hitting the network and stuff. Plus we use this helper wrapping function that may also mask things from the actual method we wanted to test.

Reply:
Multiple things here: 1. you're right I should do more fine-grained tests of the important functions below, will add them 2. hitting the network: unavoidable since the code path for downloading is quite different so this needs to be tested :woman_shrugging: and it can't really be mocked well because you'd need to mock `pip` download things, that seems more trouble than it's worth (open for ideas though), 3. yea you're right @ wrapping, my tests don't go deep enough

will add more tests tomorrow, stay tuned :+1:

Reply:
I added more tests now, feel free to protest if still any are missing but for now I think this can be resolved :+1: